core/blockchain.go

```diff
@@ -152,6 +152,11 @@ func NewBlockChain(chainDb ethdb.Database, config *params.ChainConfig, pow pow.P
 	// Check the current state of the block hashes and make sure that we do not have any of the bad blocks in our chain
 	for hash, _ := range BadHashes {
 		if header := bc.GetHeaderByHash(hash); header != nil {
+			// make sure the headerByNumber (if present) is in our current canonical chain
+			if headerByNumber := bc.GetHeaderByNumber(header.Number.Uint64()); headerByNumber != nil && headerByNumber.Hash() != header.Hash() {
```

Comment:
This check should be either of:

 * `if headerByNumber != nil && headerByNumber.Hash() == header.Hash() { rewind }`
 * `if headerByNumber == nil || headerByNumber.Hash() != header.Hash() { continue } rewind`
 
Although in theory the current implementation should also work, the conditions are imho not the cleanest combo :)